fuel-vm/src/interpreter/executors/main.rs

```diff
@@ -134,6 +163,98 @@ impl<T> Interpreter<PredicateStorage, T> {
         Ok(())
     }
 
+    async fn verify_predicate_async<Tx, E>(
+        tx: Tx,
+        balances: InitialBalances,
+        params: ConsensusParameters,
+        gas_costs: GasCosts,
+    ) -> Result<PredicatesChecked, PredicateVerificationFailed>
+    where
+        Tx: ExecutableTransaction + Send + 'static,
+        E: ParallelExecutor
+            + ParallelExecutor<TaskResult = Result<Word, PredicateVerificationFailed>>,
+    {
+        if !tx.check_predicate_owners(&params.chain_id) {
+            return Err(PredicateVerificationFailed::InvalidOwner)
+        }
+
+        let mut verifications = vec![];
+
+        for index in 0..tx.inputs().len() {
+            let is_predicate = matches!(
+                tx.inputs()[index],
+                Input::CoinPredicate(_)
+                    | Input::MessageCoinPredicate(_)
+                    | Input::MessageDataPredicate(_)
+            );
+
+            if !is_predicate {
+                continue
+            }
+
+            if let Some(predicate) = RuntimePredicate::from_tx(&params, &tx, index) {
+                let gas_costs = gas_costs.clone();
+                let balances = balances.clone();
+
+                let tx = tx.clone();
+
+                let verify_task = E::create_task(move || {
+                    let mut vm = Interpreter::with_storage(
+                        PredicateStorage::default(),
+                        params,
+                        gas_costs,
+                    );
+
+                    let context = Context::PredicateVerification { program: predicate };
+
+                    let available_gas =
+                        if let Some(x) = tx.inputs()[index].predicate_gas_used() {
+                            x
+                        } else {
+                            return Err(PredicateVerificationFailed::GasNotSpecified)
+                        };
+
+                    vm.init_predicate(context, &tx, balances.clone(), available_gas)?;
```

Comment:
The `init_predicate` should work with `Tx` instead of the reference=)

Reply:
in the signature it's `tx: &Tx,` 

```rust
pub fn init_predicate(
        &mut self,
        context: Context,
        tx: &Tx,
        balances: InitialBalances,
        gas_limit: Word,
    ) -> Result<(), InterpreterError> {
    ````

Reply:
Yes, but we do a clone inside of this function. Instead, we can change the signature to require a value and remove cloning=)